### PR TITLE
Add validation to licence fields

### PR DESCRIPTION
**Pull request by 1pretz1**
Adds validation rules as specified in the [Trello card](https://trello.com/c/uf3E7dCl/1974-add-validation-to-new-licence-document-fields). 

More info in the commits ->

## Before

Screenshot 2023-06-22 at 18 13 58

## After

Screenshot 2023-06-26 at 15 22 25
Screenshot 2023-06-26 at 15 16 48
Screenshot 2023-06-26 at 15 17 45
Screenshot 2023-06-26 at 15 17 59
Screenshot 2023-06-26 at 15 19 21
Screenshot 2023-06-26 at 15 19 45
Screenshot 2023-06-26 at 15 20 14

Trello:
https://trello.com/c/uf3E7dCl/1974-add-validation-to-new-licence-document-fields

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

File: app/validators/licence_identifier_unique_validator.rb
Comment (KludgeKML) [inline app/validators/licence_identifier_unique_validator.rb:7]: How fast is this? Is it making 309 api calls every time we try to save a licence?

**Review by KludgeKML — APPROVED** [review]
LGTM (Pending content approval of warning messages)

File: app/validators/licence_identifier_unique_validator.rb
Comment (KludgeKML) [reply app/validators/licence_identifier_unique_validator.rb:7]: (Update - it's more like 8 API calls, as they're done in batches of 50, and there isn't currently a good optimisation we can do here. Appears to be quick in integration, we should just accept this is not perfect but necessary and keep an eye on whether it causes slowdowns in use. Marking as resolved).

File: app/views/metadata_fields/_licences.html.erb
Comment (MartinJJones) [inline app/views/metadata_fields/_licences.html.erb]: Small thing, the browser will likely fix this anyway when rendering the page, but the opening tag here is a `h2`, and the closing tag is a `h3`.

**Review by MartinJJones — COMMENTED** [review]
Changes look good to me overall, just one small thing to fix 👍

File: app/views/metadata_fields/_licences.html.erb
Comment (1pretz1) [reply app/views/metadata_fields/_licences.html.erb]: Good spot, now fixed!

File: app/validators/link_or_identifier_validator.rb
Comment (KludgeKML) [inline app/validators/link_or_identifier_validator.rb]: Could there be a better name for these functions? Because it reads like: trigger an error if either link and id exists, or link and id are missing.

Comment (1pretz1) [reply app/validators/link_or_identifier_validator.rb]: Sure, we can rejig the naming if it's not clear, however your assumption of 'trigger an error if either link and id exists, or link and id are missing.'  is correct. We only want to trigger an alert if:

a) the link AND identifier fields are both filled in or 
b) the link AND identifier fields are both blank. 

A user filling in one OR the other is the correct procedure.

How about, `both_link_and_identifier_exist? || both_link_and_identifier_missing?` - thoughts? Happy for other suggestions too!

Comment (KludgeKML) [reply app/validators/link_or_identifier_validator.rb]: Yeah, I like those new names. Nice.

**Review by MartinJJones — APPROVED** [review]
Frontend changes look good to me 👍

**Review by KludgeKML — APPROVED** [review]
Looks good to me (with caveat of the variable names)

File: app/validators/link_or_identifier_validator.rb
Comment (1pretz1) [reply app/validators/link_or_identifier_validator.rb]: Done!

**Pull request merged**
